### PR TITLE
Creating infrastructure for Spark 2.4 and 3.x DSv2 connectors

### PR DESCRIPTION
**Pull request by davidrabinowitz**

### CODE REVIEW

File: spark-bigquery-parent/pom.xml [hunk 1]
```diff
@@ -0,0 +1,354 @@
...
+            </manifest>
+          </archive>
+        </configuration>
+      </plugin>
+      <!--
+      <plugin>
+          <groupId>com.coveo</groupId>
+          <artifactId>fmt-maven-plugin</artifactId>
+          <version>2.9.1</version>
+          <executions>
```
Comment (himanshukohli09) [inline spark-bigquery-parent/pom.xml:220]: Please notice, commented code.

File: spark-bigquery-python-lib/pom.xml [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+  <modelVersion>4.0.0</modelVersion>
+  <parent>
+    <groupId>com.google.cloud.spark</groupId>
+    <artifactId>spark-bigquery-parent</artifactId>
+    <version>0.30.0-SNAPSHOT</version>
+    <relativePath>../spark-bigquery-parent</relativePath>
+  </parent>
+
+  <artifactId>spark-bigquery-support</artifactId>
+  <name>PySpark-BigQuery support library</name>
```
Comment (himanshukohli09) [inline spark-bigquery-python-lib/pom.xml:13]: should we add version tag to the python lib?

File: .mvn/wrapper/MavenWrapperDownloader.java [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+    private static final String MAVEN_WRAPPER_JAR_PATH =
+            ".mvn/wrapper/maven-wrapper.jar";
+
+    /**
+     * Name of the property which should be used to override the default download url for the wrapper.
+     */
+    private static final String PROPERTY_NAME_WRAPPER_URL = "wrapperUrl";
+
+    public static void main(String args[]) {
+        System.out.println("- Downloader started");
```
[the remark this answers is not in the document]
Comment (davidrabinowitz) [reply .mvn/wrapper/MavenWrapperDownloader.java:49]: It's a maven generated code, used once to download maven. Not sure if we should touch it

File: .mvn/wrapper/MavenWrapperDownloader.java [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+            System.exit(0);
+        } catch (Throwable e) {
+            System.out.println("- Error downloading");
+            e.printStackTrace();
+            System.exit(1);
+        }
+    }
+
+    private static void downloadFileFromURL(String urlString, File destination) throws Exception {
+        if (System.getenv("MVNW_USERNAME") != null && System.getenv("MVNW_PASSWORD") != null) {
```
[the remark this answers is not in the document]
Comment (davidrabinowitz) [reply .mvn/wrapper/MavenWrapperDownloader.java:98]: It's a maven generated code, used once to download maven. Not sure if we should touch it

File: .mvn/wrapper/MavenWrapperDownloader.java [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+import java.nio.channels.*;
+import java.util.Properties;
+
+public class MavenWrapperDownloader {
+
+    private static final String WRAPPER_VERSION = "0.5.6";
+    /**
+     * Default URL to download the maven-wrapper.jar from, if no 'downloadUrl' is provided.
+     */
+    private static final String DEFAULT_DOWNLOAD_URL = "https://repo.maven.apache.org/maven2/io/takari/maven-wrapper/"
```
[the remark this answers is not in the document]
Comment (davidrabinowitz) [reply .mvn/wrapper/MavenWrapperDownloader.java:27]: Why? In any case - It's a maven generated code, used once to download maven. Not sure if we should touch it

File: mvnw [hunk 6]
```diff
@@ -0,0 +1,310 @@
+#!/bin/sh
```
[the remark this answers is not in the document]
Comment (davidrabinowitz) [reply mvnw:1]: `#!/bin/sh` works on Linux as well, for Windows there's mvnw.cmd

File: mvnw [hunk 7]
```diff
@@ -0,0 +1,310 @@
...
+  echo "Error: JAVA_HOME is not defined correctly." >&2
+  echo "  We cannot execute $JAVACMD" >&2
+  exit 1
+fi
+
+if [ -z "$JAVA_HOME" ] ; then
+  echo "Warning: JAVA_HOME environment variable is not set."
+fi
+
+CLASSWORLDS_LAUNCHER=org.codehaus.plexus.classworlds.launcher.Launcher
```
[the remark this answers is not in the document]
Comment (davidrabinowitz) [reply mvnw:162]: It's a maven generated code. Not sure if we should touch it

File: spark-bigquery-python-lib/pom.xml [hunk 8]
```diff
@@ -0,0 +1,35 @@
+<project xmlns="http://maven.apache.org/POM/4.0.0"
+  xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+  xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+  <modelVersion>4.0.0</modelVersion>
+  <parent>
```
[the remark this answers is not in the document]
Comment (davidrabinowitz) [reply spark-bigquery-python-lib/pom.xml:5]: The python lib has a parent configured, see spark-bigquery-python-lib/pom.xml

File: spark-bigquery-dsv1/spark-bigquery-with-dependencies_2.11/pom.xml [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+  xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+  <modelVersion>4.0.0</modelVersion>
+  <parent>
+    <groupId>com.google.cloud.spark</groupId>
+    <artifactId>spark-bigquery-with-dependencies-parent</artifactId>
+    <version>0.30.0-SNAPSHOT</version>
+    <relativePath>../spark-bigquery-with-dependencies-parent</relativePath>
+  </parent>
+
+  <artifactId>spark-bigquery-with-dependencies_2.11</artifactId>
```
[the remark this answers is not in the document]
Comment (davidrabinowitz) [reply spark-bigquery-dsv1/spark-bigquery-with-dependencies_2.11/pom.xml:12]: No - they take it from their parents

File: bigquery-connector-common/src/test/java/com/google/cloud/bigquery/connector/common/ParallelArrowReaderTest.java [hunk 10]
```diff
@@ -84,76 +79,64 @@ ArrowReader getReaderWithSequence(int... values) throws IOException {
...
-        read.add(((IntVector) root.getVector(0)).get(0));
-      }
-      reader.close();
-    }
-
-    assertThat(read).containsExactlyElementsIn(ImmutableList.of(0)).inOrder();
-    assertThat(executor.isShutdown()).isTrue();
-  }
+//  @Test
+//  public void testReadsAllBatchesInRoundRobin() throws Exception {
```
[the remark this answers is not in the document]
Comment (davidrabinowitz) [reply bigquery-connector-common/src/test/java/com/google/cloud/bigquery/connector/common/ParallelArrowReaderTest.java:83]: I know, this test is problematic for some reason. I want to get back to it once the refactor is over

File: bigquery-connector-common/pom.xml [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+  <name>BigQuery Connector Common Library</name>
+  <licenses>
+    <license>
+      <name>Apache License, Version 2.0</name>
+      <url>http://www.apache.org/licenses/LICENSE-2.0.txt</url>
+      <distribution>repo</distribution>
+    </license>
+  </licenses>
+  <dependencies>
+<!--    <dependency>-->
```
[the remark this answers is not in the document]
Comment (davidrabinowitz) [reply bigquery-connector-common/pom.xml:22]: Done

File: spark-bigquery-dsv1/src/test/scala/com/google/cloud/spark/bigquery/it/SparkBigQueryEndToEndReadITSuite.scala [hunk 12]
```diff
@@ -259,22 +258,23 @@ class SparkBigQueryEndToEndReadITSuite extends FunSuite
...
-//        .option("readDataFormat", "arrow")
-//        .load().cache()
-//
-//      assert(df.head() == allTypesTable.head())
-//
-//      // read from cache
-//      assert(df.head() == allTypesTable.head())
-//      assert(df.schema == allTypesTable.schema)
-//    }
+    //    test("cache data frame in DataSource %s. Data Format %s"
```
[the remark this answers is not in the document]
Comment (davidrabinowitz) [reply spark-bigquery-dsv1/src/test/scala/com/google/cloud/spark/bigquery/it/SparkBigQueryEndToEndReadITSuite.scala:261]: All integrations tests are converted to Java so they can be used in multiple spark/scala versions

File: spark-bigquery-python-lib/pom.xml (hunk 2 above)
Comment (davidrabinowitz) [reply spark-bigquery-python-lib/pom.xml:13]: It gets it from its parent. See installed version:
```
$ ls -l ~/.m2/repository/com/google/cloud/spark/spark-bigquery-support/0.30.0-SNAPSHOT/
total 16
-rw-r--r-- 1 *** ***  230 Jul 18 22:17 _remote.repositories
-rw-r--r-- 1 *** ***  733 Jul 18 22:17 maven-metadata-local.xml
-rw-r--r-- 1 *** *** 1049 Jul  9 19:14 spark-bigquery-support-0.30.0-SNAPSHOT.pom
-rw-r--r-- 1 *** *** 1684 Jul 18 22:17 spark-bigquery-support-0.30.0-SNAPSHOT.zip
```

File: spark-bigquery-parent/pom.xml (hunk 1 above)
Comment (davidrabinowitz) [reply spark-bigquery-parent/pom.xml:220]: Done

File: .mvn/wrapper/MavenWrapperDownloader.java [hunk 13]
```diff
@@ -0,0 +1,117 @@
...
+                url = mavenWrapperProperties.getProperty(PROPERTY_NAME_WRAPPER_URL, url);
+            } catch (IOException e) {
+                System.out.println("- ERROR loading '" + MAVEN_WRAPPER_PROPERTIES_PATH + "'");
+            } finally {
+                try {
+                    if(mavenWrapperPropertyFileInputStream != null) {
+                        mavenWrapperPropertyFileInputStream.close();
+                    }
+                } catch (IOException e) {
+                    // Ignore ...
```
[the remark this answers is not in the document]
Comment (davidrabinowitz) [reply .mvn/wrapper/MavenWrapperDownloader.java:72]: It's a maven generated code, used once to download maven. Not sure if we should touch it

File: spark-bigquery-dsv1/spark-bigquery-with-dependencies_2.11/pom.xml (hunk 9 above)
[the remark this answers is not in the document]
Comment (himanshukohli09) [reply spark-bigquery-dsv1/spark-bigquery-with-dependencies_2.11/pom.xml:12]: got it

**Review by himanshukohli09 — APPROVED** [review]

**Pull request merged**